### PR TITLE
Feat (Core): Implement abandoned job detection and recovery

### LINKED ISSUE #30367: Implement Abandoned Job Detection and Recovery
### Parent Issue

https://github.com/dotCMS/core/issues/29474

### Task

We need to enhance our job queue system to handle abandoned jobs. These are jobs that may have been interrupted due to server crashes, network failures, or other unexpected issues, leaving them in an inconsistent state.

#### Objective:
Implement mechanisms to detect abandoned jobs and provide recovery strategies to ensure system reliability and data consistency.

Proposed Strategies:

1. Job Heartbeats:
   - Implement periodic heartbeat updates for running jobs
   - Create a background process to identify jobs with stale heartbeats

2. Timeout Mechanisms:
   - Add a `max_execution_time` field to job configurations
   - Implement a background process to check for jobs exceeding their maximum execution time

3. Recovery Procedures:
   - Develop a recovery process
   - Identify jobs in inconsistent states and apply appropriate recovery actions

### Additional Considerations:
- Ensure that abandoned job recovery doesn't conflict with distributed locking mechanisms
- Consider the impact on job queue performance and optimize where necessary
- Evaluate and document any changes to the system's fault tolerance and high availability characteristics

### Proposed Objective

Core Features

### Proposed Priority

Priority 2 - Important

### Acceptance Criteria

1. System can detect jobs that have been abandoned due to server crashes or other issues
2. Abandoned jobs are automatically handled according to configured recovery strategies
3. All new functionality is covered by appropriate tests
4. System performance is not significantly impacted by new abandoned job handling processes

### LINKED ISSUE #30668: Issue: Job Processing Loop Fails Due to Database Connection Timeout
### Parent Issue

https://github.com/dotCMS/core/issues/29498

### Problem Statement

The job processing loop in JobQueueManagerAPIImpl is failing intermittently with a DotDataException caused by a database connection timeout. This results in an inability to fetch the next job for processing and potentially disrupts job queue management. The error logs indicate the following stack trace:

```
23:43:06.004  ERROR api.JobQueueManagerAPIImpl - Unexpected error in job processing loop: Error fetching next job
com.dotmarketing.exception.DotDataException: Error fetching next job
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processNextJob(JobQueueManagerAPIImpl.java:613) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processJobs(JobQueueManagerAPIImpl.java:573) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.lambda$start$0(JobQueueManagerAPIImpl.java:217) ~[?:?]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317) ~[?:?]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: com.dotcms.jobs.business.queue.error.JobQueueDataException: Database error while fetching next job
	at com.dotcms.jobs.business.queue.PostgresJobQueue.nextJob(PostgresJobQueue.java:573) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processNextJob(JobQueueManagerAPIImpl.java:604) ~[?:?]
	... 8 more
Caused by: com.dotmarketing.exception.DotDataException: jdbc/dotCMSPool - Connection is not available, request timed out after 5991ms.{
  "SQL": ["UPDATE job_queue SET state = ? WHERE id = (SELECT id FROM job_queue WHERE state = ? ORDER BY priority DESC, created_at ASC LIMIT 1 FOR UPDATE SKIP LOCKED) RETURNING *"],
  "maxRows": [-1],
  "offest": [0],
  "params": [
    "RUNNING",
    "PENDING"
  ]
}
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:310) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadObjectResults(DotConnect.java:997) ~[?:?]
	at com.dotcms.jobs.business.queue.PostgresJobQueue.nextJob(PostgresJobQueue.java:562) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processNextJob(JobQueueManagerAPIImpl.java:604) ~[?:?]
	... 8 more
Caused by: com.dotmarketing.exception.DotRuntimeException: jdbc/dotCMSPool - Connection is not available, request timed out after 5991ms.
	at com.dotmarketing.db.DbConnectionFactory.getConnection(DbConnectionFactory.java:236) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:599) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:308) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadObjectResults(DotConnect.java:997) ~[?:?]
	at com.dotcms.jobs.business.queue.PostgresJobQueue.nextJob(PostgresJobQueue.java:562) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processNextJob(JobQueueManagerAPIImpl.java:604) ~[?:?]
	... 8 more
Caused by: java.sql.SQLTransientConnectionException: jdbc/dotCMSPool - Connection is not available, request timed out after 5991ms.
	at com.zaxxer.hikari.pool.HikariPool.createTimeoutException(HikariPool.java:689) ~[?:?]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:196) ~[?:?]
	at com.zaxxer.hikari.pool.HikariPool.getConnection(HikariPool.java:161) ~[?:?]
	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:100) ~[?:?]
	at com.dotmarketing.db.DbConnectionFactory.getConnection(DbConnectionFactory.java:223) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:599) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:308) ~[?:?]
	at com.dotmarketing.common.db.DotConnect.loadObjectResults(DotConnect.java:997) ~[?:?]
	at com.dotcms.jobs.business.queue.PostgresJobQueue.nextJob(PostgresJobQueue.java:562) ~[?:?]
	at com.dotcms.jobs.business.api.JobQueueManagerAPIImpl.processNextJob(JobQueueManagerAPIImpl.java:604) ~[?:?]
	... 8 more
23:43:06.004  ERROR license.LicenseManager - Could not detect if server 5b7f00e4-49c5-48a0-8bd5-712cbfd0153c is duplicated
com.dotmarketing.exception.DotDataException: jdbc/dotCMSPool - Connection is not available, request timed out after 5991ms.{
  "SQL": ["SELECT id FROM sitelic WHERE (serverid = ? OR license = ?) AND startup_time != ?"],
  "maxRows": [-1],
  "offest": [0],
  "params": [
    "5b7f00e4-49c5-48a0-8bd5-712cbfd0153c",
    "0toaiuQe4vxQiX/5ewQQj9N6JGBqmb3WUdbMBmCUC17vYK86AX3HJ6PZoNI70BfRKzMBhRurLeICbPxZIOYhrwJKk7lgT39564Rt56FDOZAhvEidREV3YKZZrdffGGgF1sLGilVWKlU69i0wEoCD+SbtmT5DEdFamklkfvUtyKMAAAAIRGV2IFRlc3QAAAAIAAABWbfTJtkAAAAIAAAAAAAAAAAAAAAEAAAB9AAAACRjNmVhYzkzZS0wOGNiLTRjOTItYmFlMi05NmNmYjY4Y2ZjMDMAAAAEcHJvZAAAAAQAAAABAAAABAAAAAAAAAAEAAABLAAAACQxNzUwOTlkMS1iMzM1LTRmYzYtOTYzYS0wMGM5YTdhYTcyYzQ=",
    1731626407189
  ]
}

```

 This problem arose with very little load 

### Steps to Reproduce

Unfortunately, I do not have a pattern to reproduce this 
It started happening on a long-running instance. Like the next day

### Acceptance Criteria

We need to revise the Database connection pool configuration (dotCMSPool) and determine if any different configuration can resolve the problem or ensure recovery. Perhaps a dedicated pool or something. 

### dotCMS Version

main

### Proposed Objective

Technical User Experience

### Proposed Priority

Priority 2 - Important

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by jgambarios**
This pull request includes some refactoring and improvements to the job queue management system. 
The changes primarily focus on removing the job polling mechanism, enhancing job validation, and adding an abandoned job detector. Below are the most important changes:

### Removal of Job Polling Mechanism:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueConfig.java`](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL13-L24): Removed the `pollJobUpdatesIntervalMilliseconds` parameter and its associated methods. [[1]](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL13-L24) [[2]](diffhunk://#diff-4c8b70e959cc00105adc8298bf8a977a3c18a6ceb5d55110bda6d023159a5e5aL36-L44)
* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueConfigProducer.java`](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L19-L23): Removed the default polling interval configuration and its usage. [[1]](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L19-L23) [[2]](diffhunk://#diff-1367d122323650421a37296f7a223becc19bb3d6c1d2a965dd52ae6bbd2f5e74L33-R28)
* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L121-R122): Removed the polling scheduler and related methods. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L121-R122) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L477-L498)

### Addition of Abandoned Job Detector:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R112): Added `AbandonedJobDetector` as a dependency and integrated it into the job queue lifecycle methods. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R112) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L142-R151) [[3]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L170-L179) [[4]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L221-R209) [[5]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R230-L243)

### Enhancements to Job Validation:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303): Added validation for job parameters if the processor implements `Validator` and handled validation exceptions.

### Event Handling Improvements:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303): Replaced direct event firing with `JobUtil.sendEvents` for better event management. [[1]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561R283-R303) [[2]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L543-R526) [[3]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L668-R649) [[4]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L721-R703) [[5]](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L743-R724)

### Miscellaneous Changes:

* [`dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java`](diffhunk://#diff-c092f8af2f800c0ca84f2c2f4aed0af60d0ed488e20fa32992967e154064f561L764-R743): Adjusted the job progress update interval from 2 to 3 seconds.

This PR fixes: #30367

### CODE REVIEW

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/jobs/business/queue/PostgresJobQueue.java [hunk 1]
```diff
@@ -580,6 +618,60 @@ public Job nextJob() throws JobQueueDataException, JobLockingException {
...
+                        .build();
+                final JobResult jobResult = JobResult.builder().errorDetail(errorDetail).build();
+
+                final Job abandonedJob = foundAbandonedJob.markAsAbandoned(jobResult);
+                updateJobStatus(abandonedJob);
+
+                return abandonedJob;
+            }
+
+            return null;
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/jobs/business/queue/PostgresJobQueue.java]: Done

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/job/JobQueueResource.java [hunk 2]
```diff
@@ -54,37 +57,50 @@ public JobQueueResource(WebResource webResource, JobQueueHelper helper,
...
+            return ExceptionMapperUtil.createResponse(null, e.getMessage());
+        }
     }
 
     @POST
     @Path("/{queueName}")
     @Consumes(MediaType.APPLICATION_JSON)
     @Produces(MediaType.APPLICATION_JSON)
-    public ResponseEntityView<String> createJob(
+    public Response createJob(
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/job/JobQueueResource.java:85]: Now we can validate the job before creation, and if the job fails when validating, the exception needs to be handled:

```java
        } catch (JobValidationException e) {
            return ExceptionMapperUtil.createResponse(null, e.getMessage());
        }
```

The `ExceptionMapperUtil.createResponse` returns a `Response`, that's why I made the change.

File: dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    static final int DEFAULT_JOB_ABANDONMENT_THRESHOLD_MINUTES = Config.getIntProperty(
+            "JOB_ABANDONMENT_THRESHOLD_MINUTES", 30
+    );
+
+    /**
+     * Produces a configuration object for the abandoned job detector.
+     *
+     * @return A new configuration object with the current property values
+     */
+    @Produces
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java:41]: Thanks, this is important to know.
Done.

File: dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * abandoned jobs and the threshold for considering a job abandoned.
+ * <p>
+ * Configuration is read from the following properties:
+ * <ul>
+ *   <li>JOB_ABANDONMENT_DETECTION_INTERVAL_MINUTES: How often to check for abandoned jobs (default: 5)</li>
+ *   <li>JOB_ABANDONMENT_THRESHOLD_MINUTES: How long before a non-updating job is considered abandoned (default: 30)</li>
+ * </ul>
+ */
+@ApplicationScoped
+public class AbandonedJobDetectorConfigProducer {
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/jobs/business/detector/AbandonedJobDetectorConfigProducer.java:19]: Done.

File: dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java [hunk 5]
```diff
@@ -139,24 +136,19 @@ public class JobQueueManagerAPIImpl implements JobQueueManagerAPI {
...
+     * @param jobProcessorFactory  The JobProcessorFactory for creating job processors instances.
+     * @param retryPolicyProcessor The RetryPolicyProcessor for processing retry policies.
+     * @param abandonedJobDetector The AbandonedJobDetector for detecting abandoned jobs.
      */
     @Inject
     public JobQueueManagerAPIImpl(@Named("queueProducer") JobQueue jobQueue,
             JobQueueConfig jobQueueConfig,
             CircuitBreaker circuitBreaker,
             @DefaultRetryStrategy RetryStrategy defaultRetryStrategy,
             RealTimeJobMonitor realTimeJobMonitor,
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java:148]: We can have multiple `RetryStrategy` implementation as each processor could have its own retry strategy, but if no retry strategy is provided for a processor we need to use the default, and the default retry strategy is produced in the `RetryStrategyProducer`.

File: dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java [hunk 6]
```diff
@@ -218,7 +202,11 @@ public void start() {
                 });
             }
 
-            Logger.info(this, "JobQueue has been successfully started.");
+            // Start the abandoned job detector
+            abandonedJobDetector.start();
+
+            Logger.info(this,
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java:208]: is this going to log every 5 mins? couldn't it be logging too much?

Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/jobs/business/api/JobQueueManagerAPIImpl.java:208]: No, this is logged only when the `JobQueueManagerAPIImpl` starts, should be just once.

**Review by nollymar — APPROVED** [review]

**Pull request merged**